core/schema/secret.go

```diff
@@ -49,27 +63,59 @@ type setSecretArgs struct {
 
 func (s *secretSchema) setSecret(ctx context.Context, parent *core.Query, args setSecretArgs) (dagql.Instance[*core.Secret], error) {
 	var inst dagql.Instance[*core.Secret]
-	if err := parent.Secrets.AddSecret(ctx, args.Name, []byte(args.Plaintext)); err != nil {
+
+	accessor, err := s.makeAccessor(ctx, parent, args.Name)
+	if err != nil {
+		return inst, err
+	}
+
+	if err := parent.Secrets.AddSecret(ctx, accessor, []byte(args.Plaintext)); err != nil {
 		return inst, err
 	}
 	// NB: to avoid putting the plaintext value in the graph, return a freshly
 	// minted Object that just gets the secret by name
 	if err := s.srv.Select(ctx, s.srv.Root(), &inst, dagql.Selector{
 		Field: "secret",
 		Args: []dagql.NamedInput{
-			{Name: "name", Value: dagql.NewString(args.Name)},
+			{
+				Name:  "name",
+				Value: dagql.NewString(args.Name),
+			},
+			{
+				Name:  "accessor",
+				Value: dagql.Opt(dagql.NewString(accessor)),
+			},
 		},
 	}); err != nil {
 		return inst, err
 	}
+
 	return inst, nil
 }
 
-func (s *secretSchema) plaintext(ctx context.Context, parent *core.Secret, args struct{}) (dagql.String, error) {
-	bytes, err := parent.Plaintext(ctx)
+func (s *secretSchema) plaintext(ctx context.Context, secret *core.Secret, args struct{}) (dagql.String, error) {
+	bytes, err := secret.Query.Secrets.GetSecret(ctx, secret.Accessor)
 	if err != nil {
+		if errors.Is(err, secrets.ErrNotFound) {
+			return "", errors.Wrapf(secrets.ErrNotFound, "secret %s", secret.Name)
+		}
 		return "", err
 	}
 
 	return dagql.NewString(string(bytes)), nil
 }
+
+func (s *secretSchema) makeAccessor(ctx context.Context, parent *core.Query, name string) (string, error) {
+	m, err := parent.CurrentModule(ctx)
+	if err != nil && !errors.Is(err, core.ErrNoCurrentModule) {
+		return "", err
+	}
+	var d digest.Digest
+	if m != nil {
+		d, err = m.InstanceID.Digest()
+		if err != nil {
+			return "", err
+		}
```

Comment:
:eyes: so this seems to work like a stable ID per module...

@vito does it seem reasonable that this would stay somewhat stable between runs? If not, it doesn't have sufficient uniqueness guarantees, I can go and add a module ID somewhere, but I'd like to avoid doing all the extra special casing for that if possible.

Reply:
You should use `m.Source.ID().Digest()`, I setup source to be usable as a stable ID with the exact properties you described [here](https://github.com/dagger/dagger/pull/6601#issuecomment-1954348045). Was previously needed to detecting dependency loops, but that got refactored away recently, but I guess it will come in handy again here.

Reply:
*perfect*, awesome! Thanks, have updated to use that as well.